### PR TITLE
Navigate and edit cells within a range

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR adds a new feature that allows users to navigate across cells within the selected range, like in other Spreadsheets in the market with <kbd>Enter</kbd> and <kbd>Tab</kbd> keys. It helps in the quick editing of data.

If two or more ranges are selected, the navigation happens only within the range containing focus (focused cell). When the selection contains headers, the focus jumps to the nearest cell and continues traversing only within the cell range.

### Navigation
- <kbd>Enter</kbd> moves down
- <kbd>Shift + Enter</kbd> moves up
- <kbd>Tab</kbd> moves right
- <kbd>Shift + Tab</kbd> moves left

The selection wraps similarly to enabled `autoWrapCol` and `autoWrapRow` and cannot be changed.

### Editing data

To open the editor, start typing by pressing any alphanumeric key meaning letters (A-Z), numeral keys (0-9), additional symbols, and special characters. When the editor is opened:

- <kbd>Enter</kbd> saves data
- <kbd>Escape</kbd> closes the editor without saving

#### Simple demo
Kapture 2024-02-08 at 15 17 11

https://jsfiddle.net/4r9Lheou/

### Support for focus navigation within a range that contains merged cells 

The PR adds support for focus navigation for merged cells. From now on, the focus jumps between merged cells and does not navigate through the hidden cells that these merged cells cover. The behavior only applies when the navigation within a range is performed.

Kapture 2024-03-18 at 14 11 13

### Other changes
 * Add a new `isHeader` method to the _CellRange_ class;
 * Add a new `beforeSelectionFocusSet` hook that is triggered before changing the position of the focused cell;
 * Add a new `afterSelectionFocusSet` hook that is triggered after changing the position of the focused cell;
 * Add a new `modifyTransformFocus` hook that is triggered before the cell transformation is applied to the focused cell;
 * Add a new `afterModifyTransformFocus` hook that is triggered after the cell transformation is applied to the focused cell;
 * Fix wrong cell selection of the merge cells when the selection was done from bottom-right to top-left;

**Things to do:**
- [x] Add a logic that allows navigating the cell/focus within a range;
  - [x] Add new hooks, if necessary;
- [x] Add a logic that allows navigating the cell/focus within a range with merged cells;
- [x] ~Add a logic that allows navigating the cell/focus within a range with nested headers + collapsible columns;~ (the focus should not be involved in navigation when headers are selected, [thread](https://handsoncode.slack.com/archives/C4PV4S33Q/p1707212381888479))
- [x] Add tests for the `tabMoves` options;
- [x] Add tests for the `enterMoves` options;
- [x] Add tests for new roles of the keyboard shortcuts; 
  - [x] <kbd>Enter</kbd> moves down
  - [x] <kbd>Shift + Enter</kbd> moves up
  - [x] <kbd>Tab</kbd> moves right
  - [x] <kbd>Shift + Tab</kbd> moves left
- [x] Add a logic that allows expand the selection correctly when the focus is moved around;

### How has this been tested?

I tested the changes locally and I covered the stuff with tons of new tests.

### Types of changes

- [x] New feature or improvement (non-breaking change which adds functionality)

### Related issue(s):
1. https://github.com/handsontable/dev-handsontable/issues/1690
2. https://github.com/handsontable/dev-handsontable/issues/74
3. https://github.com/handsontable/dev-handsontable/issues/1344
4. https://github.com/handsontable/dev-handsontable/issues/1148

### Affected project(s):
- [x] `handsontable`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

### CODE REVIEW

**Comment by krzysztofspilka** [comment]
Issue 1: I got into a loop when trying to move the selector with <kbd>TAB</kbd> at the edge of the merge cell. Also, if the column selected is adjacent to a merged cell, the selection is stuck.

https://github.com/handsontable/handsontable/assets/8048526/cb9c5355-6d1b-44b7-874e-8167251670af

https://github.com/handsontable/handsontable/assets/8048526/59740009-9485-4c06-a191-923739d39050

**Comment by krzysztofspilka** [comment]
Issue 2. To select an item from the select cell type, I first focus on a cell and press <kbd>Space</kbd> twice. Then I move through the menu using <kbd>arrow keys</kbd> and then press <kbd>Enter</kbd> to choose the right value.

Unfortunately, it loses the focus, so when I press <kbd>Tab</kbd>, I end up in the `input` below the grid.

https://github.com/handsontable/handsontable/assets/8048526/a2461c6b-bc92-4c92-8d6c-ab3abca58e4f

**Comment by krzysztofspilka** [comment]
Issue 3. Connected with issue 1 - the merged cell gets into a loop if you use <kbd>Tab</kbd> key to navigate.

https://github.com/handsontable/handsontable/assets/8048526/2e7f1a40-cd5c-4764-8279-aba34f06a5f3

**Comment by krzysztofspilka** [comment]
Issue 5. The selection is overlapped by the bottom fixed row if it's close to the bottom edge of the viewport. I moved the selection with <kbd>Tab</kbd> on this video.

https://github.com/handsontable/handsontable/assets/8048526/0f3dae82-7990-4cf3-8c3a-76479a4535d2

**Comment by krzysztofspilka** [comment]
Issue 6. The selected cells get detached from the focused cell. 

1. Select three or more rows in the first column (with mouse or however).
2. Press <kbd>Tab</kbd> twice
3. Press Shift + ↘︎ (End key)

https://github.com/handsontable/handsontable/assets/8048526/c784ac81-ccd4-48bb-aa05-2704b83a9f98

**Comment by krzysztofspilka** [comment]
Issue 7. If `tabNavigation={false}` then when I select cells and press <kbd>Tab</kbd> once, then I lose the focus. It's restored after I press that key one more time. Haven't noticed this problem with <kbd>Shift + Tab</kbd>.

https://github.com/handsontable/handsontable/assets/8048526/89618186-ecd8-4d7e-a7af-99fa223f0098

**Comment by krzysztofspilka** [comment]
- Issue 1: related, to be fixed now
- Issue 2: not related, moved to https://github.com/handsontable/dev-handsontable/issues/1746
- Issue 3: related, to be fixed now
- Issue 4: NOT related but to be fixed now https://github.com/handsontable/dev-handsontable/issues/1747
- Issue 5: not related, moved to https://github.com/handsontable/dev-handsontable/issues/1748
- Issue 6: related, to be fixed now
- Issue 7: Problem with CodeSandbox - to be confirmed

**Review by jansiegel — CHANGES_REQUESTED** [review]

File: handsontable/src/selection/highlight/visualSelection.js [hunk 1]
```diff
@@ -176,11 +176,13 @@ class VisualSelection extends Selection {
         this.cellRange = this.settings.createCellRange(cellCoordsRenderable);
       }
 
+      // TODO
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/selection/highlight/visualSelection.js:179]: It is supposed to stay. The code that it marks `broaderCellRange.setHighlight(currentHighlight);` is a workaround and should be moved away from the VisualSelection class in the future.

File: handsontable/src/pluginHooks.js [hunk 2]
```diff
@@ -743,6 +745,47 @@ const REGISTERED_HOOKS = [
    */
   'afterSelectionEndByProp',
 
+  /**
+   * Fired after the focus position within a selected range is changed.
+   *
+   * @since 14.2.0
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/pluginHooks.js]: Thanks ... fixed within 92de45b0e8b2c4732af7ed9bf009536dacfcbb01.

**Review by jansiegel — APPROVED** [review]

**Comment by krzysztofspilka** [comment]
@budnix one thing I noticed that I no longer can use `DELETE` or `BACKSPACE` to remove the content.

1. Open https://codesandbox.io/p/sandbox/vigorous-bouman-5vv2f9?file=%2Fpackage.json
2. Click on any cell
3. Press Delete or Backspace

Result: nothing happens

**Comment by budnix** [comment]
> @budnix Another thing I noticed is that I cannot use `TAB` to leave the cell in the `empty` column. I'm not sure if it's specific for this [demo](https://codesandbox.io/p/sandbox/vigorous-bouman-5vv2f9?file=%2Fpackage.json) or sth more global.

Solved by adding `autoWrapRow` and `autoWrapCol` https://codesandbox.io/p/sandbox/sparkling-cloud-d9yf4s?file=%2Fsrc%2Findex.tsx%3A173%2C20

**Pull request merged**